guides/custom_endpoints.md

```diff
@@ -0,0 +1,288 @@
+* `Membrane.File.Sink` - used for saving serialized data into a file 
+
+For each incoming track we will create separate serializer and sink.
+
+## Subscribing for a track
+
+Whenever some endpoint publishes its tracks, all other endpoints receive message
+in form of `{:new_tracks, tracks}` where `tracks` is a list of `t:Membrane.RTC.Engine.Track.t/0`.
+
+To subscribe for any of published tracks endpoint has to call `Membrane.RTC.Engine.subscribe/4`. 
+
+After subscribing for a track an endpoint will be notified about its readiness in
+`c:Membrane.Bin.handle_pad_added/3` callback. An example implementation of `handle_pad_added`
+callback can look like this
+
+```elixir
+@impl true
+def handle_pad_added(Pad.ref(:input, track_id) = pad, _ctx, state) do
+  children = %{
+    {:track_receiver, track_id} => %TrackReceiver{
+      track: Map.fetch!(state.tracks, track_id),
+      initial_target_variant: :high
+    },
+    {:serializer, track_id} => Membrane.Stream.Serializer,
+    {:sink, track_id} => %Membrane.File.Sink{
+      location: Path.join(state.output_dir_path, track_id)
+    }
+  }
+
+  links = [
+    link_bin_input(pad)
+    |> to({:track_receiver, track_id})
+    |> to({:serializer, track_id})
+    |> to({:sink, track_id})
+  ]
+
+  {{:ok, spec: %ParentSpec{children: children, links: links}}, state}
+end
+```
+
+We simply link our input pad on which we will receive media data to serializer
+and then from serializer to sink.
+
+Before serializer we also plugged in one more element called `TrackReceiver`.
+It is responsible for requesting the highest possible track variant and
+switching to other variants when the currently used one becomes inactive.
+
+> #### Deeper dive into TrackReciver {: .tip}
+>
+> After linking input pad for given track, our endpoint will start receiving events.
+> The very first event is always `TrackVariantResumed`. It informs that given track
+> variant is available and can be requested with `RequestTrackVariant` event. 
+> Engine ensures that each endpoint always receive track variant starting from
+> a keyframe therefore before receiving first media packets our endpoint 
+> will receive `TrackVariantSwitched` event and finally media packets will
+> start flowing.
+> 
+> When some track variant was paused we will receive `TrackVariantPaused` event.
+>
+> The whole process is shown in the following diagram
+> 
+> ```ascii
+> Engine ---- TrackVariantResumed (:medium) ---> Endpoint
+> Engine <--- RequestTrackVariant (:medium) ---- Endpoint
+> Engine ---- TrackVariantResumed (:low)    ---> Endpoint
+> Engine ---- TrackVariantSwitched ---> Endpoint
+> Engine ----        media         ---> Endpoint
+> Engine ---- TrackVariantPaused  (:medium) ---> Endpoint
+> Engine <--- RequestTrackVariant (:low)    ---- Endpoint
+> ```
+>
+> Beacause this logic must be duplicated in each endpoint we encapsulated 
+> it into `TrackReceiver` that can easily be plugged in before actual
+> Membrane element.
```

Comment:
```suggestion
> Because this logic must be duplicated in each endpoint we encapsulated
> it into `TrackReceiver` that can easily be plugged in before the actual
> Membrane element.
```